### PR TITLE
(tl_mlx5_6) TL/MLX5: tl stubs

### PR DESCRIPTION
**Pull request by vspetrov**
## What
Adds new TL/MLX5: minimal necessary tl iface stubs w/o much actual implementation (added in next PRs).
Adds option to provide negate sign "^" to the --with-tls. Default list of tls is set to "^mlx5" so that the new TL is not built by default. Will change that when the TL is ready.

### CODE REVIEW

File: src/components/tl/mlx5/tl_mlx5.c [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    {"DM_BUF_SIZE", "8k", "Size of the pre-allocated DeviceMemory buffer",
+     ucc_offsetof(ucc_tl_mlx5_lib_config_t, dm_buf_size),
+     UCC_CONFIG_TYPE_MEMUNITS},
+
+    {"DM_BUF_NUM", "auto", "Number of DM buffers to alloc",
+     ucc_offsetof(ucc_tl_mlx5_lib_config_t, dm_buf_num),
+     UCC_CONFIG_TYPE_MEMUNITS},
+
+    {"BLOCK_SIZE", "0",
```
Comment (Sergei-Lebedev) [inline src/components/tl/mlx5/tl_mlx5.c:29]: rename to ALLTOALL_BLOCK_SIZE if it's specific to alltoall

File: src/components/tl/mlx5/tl_mlx5.c [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+     "Size of the blocks that are sent using blocked AlltoAll Algorithm",
+     ucc_offsetof(ucc_tl_mlx5_lib_config_t, block_size), UCC_CONFIG_TYPE_UINT},
```
Comment (Sergei-Lebedev) [inline src/components/tl/mlx5/tl_mlx5.c:31]: use MEMUNTIS if size in bytes

File: src/components/tl/mlx5/tl_mlx5.c [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    {"ASR_BARRIER", "0", "Boolean - use  service barrier or p2p sync of ASRs",
+     ucc_offsetof(ucc_tl_mlx5_lib_config_t, asr_barrier), UCC_CONFIG_TYPE_UINT},
+
+    {"DM_BUF_SIZE", "8k", "Size of the pre-allocated DeviceMemory buffer",
+     ucc_offsetof(ucc_tl_mlx5_lib_config_t, dm_buf_size),
+     UCC_CONFIG_TYPE_MEMUNITS},
+
+    {"DM_BUF_NUM", "auto", "Number of DM buffers to alloc",
+     ucc_offsetof(ucc_tl_mlx5_lib_config_t, dm_buf_num),
+     UCC_CONFIG_TYPE_MEMUNITS},
```
Comment (Sergei-Lebedev) [inline src/components/tl/mlx5/tl_mlx5.c:27]: is UCC_CONFIG_TYPE_MEMUNITS is correct here?

File: src/components/tl/mlx5/tl_mlx5.c [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    {"DC_THRESHOLD", "128",
+     "If number of nodes >= DC_THRESHOLD then DC QPs "
+     "are used instead of RC",
+     ucc_offsetof(ucc_tl_mlx5_lib_config_t, dc_threshold),
+     UCC_CONFIG_TYPE_UINT},
+
+    {"DM_HOST", "0",
+     "Use host registered memory instead of DM for "
+     "transpose staging",
+     ucc_offsetof(ucc_tl_mlx5_lib_config_t, dm_host), UCC_CONFIG_TYPE_UINT},
```
Comment (Sergei-Lebedev) [inline src/components/tl/mlx5/tl_mlx5.c:46]: UCC_CONFIG_TYPE_BOOL

File: src/components/tl/mlx5/Makefile.am [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#
+# Copyright (C) Mellanox Technologies Ltd. 2022.  ALL RIGHTS RESERVED.
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/mlx5/Makefile.am:2]: done

File: src/components/tl/mlx5/Makefile.am [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	tl_mlx5_context.c \
+	tl_mlx5_team.c    \
+	tl_mlx5_coll.h
+
+module_LTLIBRARIES = libucc_tl_mlx5.la
+libucc_tl_mlx5_la_SOURCES  = $(sources)
+libucc_tl_mlx5_la_CPPFLAGS = $(AM_CPPFLAGS) $(BASE_CPPFLAGS)
+libucc_tl_mlx5_la_CFLAGS   = $(BASE_CFLAGS)
+libucc_tl_mlx5_la_LDFLAGS  = -version-info $(SOVERSION) --as-needed
+libucc_tl_mlx5_la_LIBADD   = $(UCC_TOP_BUILDDIR)/src/libucc.la -libverbs -lmlx5
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/mlx5/Makefile.am:20]: already have it in config/m4/ibverbs.m4. Switched to IBVERBS_LIBADD and MLX5DV_LIBADD here

File: src/components/tl/mlx5/configure.m4 [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#
+# Copyright (C) Mellanox Technologies Ltd. 2022.  ALL RIGHTS RESERVED.
+#
+
+tl_mlx5_enabled=n
+CHECK_TLS_REQUIRED(["mlx5"])
+AS_IF([test "$CHECKED_TL_REQUIRED" = "y"],
+[
+    mlx5_happy=no
+    if test "x$mlx5dv_happy" = "xyes" -a "x$have_mlx5dv_wr_raw_wqe" = "xyes"; then
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/mlx5/configure.m4:10]: variables defined in ibverbs.m4. First is defined if libmlx5 (and headers) are available. 2nd is there if libmlx5 is recent enough and it has the "raw wqe builder" functionality

File: src/components/tl/mlx5/tl_mlx5.c [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+ucc_status_t ucc_tl_mlx5_get_lib_attr(const ucc_base_lib_t *lib,
+                                      ucc_base_lib_attr_t * base_attr);
+ucc_status_t ucc_tl_mlx5_get_context_attr(const ucc_base_context_t *context,
+                                          ucc_base_ctx_attr_t *     base_attr);
+
+static ucc_config_field_t ucc_tl_mlx5_lib_config_table[] = {
+    {"", "", NULL, ucc_offsetof(ucc_tl_mlx5_lib_config_t, super),
+     UCC_CONFIG_TYPE_TABLE(ucc_tl_lib_config_table)},
+
+    {"ASR_BARRIER", "0", "Boolean - use  service barrier or p2p sync of ASRs",
```
[the remark this answers is not in the document]
Comment (vspetrov) [reply src/components/tl/mlx5/tl_mlx5.c:18]: ASR BARRIER is removed (variable no longer used in the algorithm)

File: src/components/tl/mlx5/tl_mlx5.c (hunk 3 above)
Comment (vspetrov) [reply src/components/tl/mlx5/tl_mlx5.c:27]: no, should be UINT
fixed

**Comment by vspetrov** [comment]
@manjugv @bureddy plz give it a look. I'd like to start moving with mlx5/a2a upstreaming

File: src/components/tl/mlx5/tl_mlx5.h [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+#include "utils/profile/ucc_profile_off.h"
+#endif
+
+#define UCC_TL_MLX5_PROFILE_FUNC          UCC_PROFILE_FUNC
+#define UCC_TL_MLX5_PROFILE_FUNC_VOID     UCC_PROFILE_FUNC_VOID
+#define UCC_TL_MLX5_PROFILE_REQUEST_NEW   UCC_PROFILE_REQUEST_NEW
+#define UCC_TL_MLX5_PROFILE_REQUEST_EVENT UCC_PROFILE_REQUEST_EVENT
+#define UCC_TL_MLX5_PROFILE_REQUEST_FREE  UCC_PROFILE_REQUEST_FREE
+
+#define DC_KEY 1
```
Comment (bureddy) [inline src/components/tl/mlx5/tl_mlx5.h:34]: is this used in this PR?

**Review by bureddy — APPROVED** [review]

**Review by Sergei-Lebedev — APPROVED** [review]

File: src/components/tl/mlx5/tl_mlx5.c (hunk 2 above)
Comment (vspetrov) [reply src/components/tl/mlx5/tl_mlx5.c:28]: no, block size is unsigned - it is number of local ranks grouped into block

File: src/components/tl/mlx5/tl_mlx5.h (hunk 9 above)
Comment (vspetrov) [reply src/components/tl/mlx5/tl_mlx5.h:34]: nope  - removed

**Pull request merged**